snuba/cli/subscriptions.py

```diff
@@ -0,0 +1,195 @@
+from snuba.subscriptions.store import RedisSubscriptionDataStore
+from snuba.subscriptions.worker import (
+    SubscriptionResultCodec,
+    SubscriptionWorker,
+)
+from snuba.util import create_metrics
+from snuba.utils.codecs import PassthroughCodec
+from snuba.utils.streams.batching import BatchingConsumer
+from snuba.utils.streams.kafka import (
+    CommitCodec,
+    KafkaConsumer,
+    KafkaProducer,
+    build_kafka_consumer_configuration,
+)
+from snuba.utils.streams.synchronized import SynchronizedConsumer
+from snuba.utils.streams.types import Topic
+
+
+@click.command()
+@click.option(
+    "--dataset",
+    "dataset_name",
+    default="events",
+    type=click.Choice(DATASET_NAMES),
+    help="The dataset to target",
+)
+@click.option("--topic")
+@click.option("--partitions", type=int)
+@click.option("--commit-log-topic")
+@click.option(
+    "--commit-log-group",
+    "commit_log_groups",
+    multiple=True,
+    default=["snuba-consumers"],
+)
+@click.option(
+    "--consumer-group",
+    default="snuba-subscriptions",
+    help="Consumer group use for consuming the dataset source topic.",
+)
+@click.option(
+    "--auto-offset-reset",
+    default="error",
+    type=click.Choice(["error", "earliest", "latest"]),
+    help="Kafka consumer auto offset reset.",
+)
+@click.option(
+    "--bootstrap-server",
+    "bootstrap_servers",
+    multiple=True,
+    help="Kafka bootstrap server(s) to use.",
+)
+@click.option(
+    "--max-batch-size",
+    default=settings.DEFAULT_MAX_BATCH_SIZE,
+    type=int,
+    help="Max number of messages to batch in memory before writing to Kafka.",
+)
+@click.option(
+    "--max-batch-time-ms",
+    default=settings.DEFAULT_MAX_BATCH_TIME_MS,
+    type=int,
+    help="Max length of time to buffer messages in memory before writing to Kafka.",
+)
+@click.option("--schedule-ttl", type=int, default=60 * 5)
+@click.option("--result-topic")
+@click.option("--log-level", default=settings.LOG_LEVEL, help="Logging level to use.")
+def subscriptions(
+    *,
+    dataset_name: str,
+    topic: Optional[str],
+    partitions: Optional[int],
+    commit_log_topic: Optional[str],
+    commit_log_groups: Sequence[str],
+    consumer_group: str,
+    auto_offset_reset: str,
+    bootstrap_servers: Sequence[str],
+    max_batch_size: int,
+    max_batch_time_ms: int,
+    schedule_ttl: int,
+    result_topic: Optional[str],
+    log_level: str,
+) -> None:
+    assert result_topic is not None
+
+    setup_logging(log_level)
+    setup_sentry()
+
+    dataset = get_dataset(dataset_name)
+
+    if not bootstrap_servers:
+        bootstrap_servers = settings.DEFAULT_DATASET_BROKERS.get(
+            dataset_name, settings.DEFAULT_BROKERS
+        )
+
+    producer = KafkaProducer(
+        {
+            "bootstrap.servers": ",".join(bootstrap_servers),
+            "partitioner": "consistent",
+            "message.max.bytes": 50000000,  # 50MB, default is 1MB
+        },
+        SubscriptionResultCodec(),
+    )
```

Comment:
https://github.com/getsentry/snuba/blob/3378bb1ec69525669669750be0d0d3311a3ae128/snuba/consumers/consumer_builder.py#L88-L94

Reply:
Are you planning to eventually move the logic in common with the consumer_builder somewhere in common, or do you think this should evolve separately ?

Reply:
I am not sure yet. I don't expect this to be the final resting place for this, but I also don't think this belongs in the consumer builder either. (I think it has too many responsibilities at this point.)

My current thinking is to add a `Cluster[TEncoded]` type to `snuba.utils.streams` that exposes `cluster.consumer(…) -> Consumer[TDecoded]` and `cluster.producer(…) -> Producer[TDecoded]`. (The type vars correspond to the [Codec](https://github.com/getsentry/snuba/blob/3378bb1ec69525669669750be0d0d3311a3ae128/snuba/utils/codecs.py#L11-L18) generic.) The configuration will probably be specified by the cluster backend at that point. Some will probably not be overridable (e.g. bootstrap servers come from the cluster definition itself, the partitioner will likely be consistent for all producers to maintain consistency with other implementations, etc.)

This is also related to https://github.com/getsentry/snuba/pull/752#discussion_r375012480.